src/Servers/Kestrel/Core/test/Http2/FlowControl/InputFlowControlTests.cs

```diff
@@ -0,0 +1,381 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.FlowControl;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Tests.Http2.FlowControl;
+
+public class InputFlowControlTests
+{
+    [Fact]
+    public async Task Threads1_Advance()
+    {
+        for (var i = 0; i < 1000; i++)
+        {
+            var sut = new InputFlowControl(5000, 1);
+            await Task.Run(() => Advance(sut)); ;
+            Assert.Throws<Http2ConnectionErrorException>(() => sut.TryAdvance(1));
+        }
+
+        static void Advance(InputFlowControl sut)
+        {
+            for (var j = 0; j < 5000; j++)
+            {
+                Assert.True(sut.TryAdvance(1));
+            }
+        }
+    }
+
+    [Fact]
+    public async Task Threads2_Advance()
+    {
+        for (var i = 0; i < 1000; i++)
+        {
+            var sut = new InputFlowControl(10000, 1);
+            var t1 = Task.Run(() => Advance(sut));
+            var t2 = Task.Run(() => Advance(sut));
+            await Task.WhenAll(t1, t2);
+            Assert.Throws<Http2ConnectionErrorException>(() => sut.TryAdvance(1));
```

Comment:
While I'm glad you did this kind of sanity testing, I'm not sure we should check this in. If we're confident something won't happen, we can verify it once rather than 1000 times each test run.

If we think these tests are valuable, I would do one iteration each per test and not waste a bunch of cycles trying to force a race condition that we're confident can't happen.